crates/env/src/api.rs

```diff
@@ -701,6 +701,106 @@ pub fn set_code_hash(code_hash: &[u8; 32]) -> Result<()> {
     <EnvInstance as OnInstance>::on_instance(|instance| instance.set_code_hash(code_hash))
 }
 
+/// **New version of this function which will replace [`set_code_hash`] in `v5.0`**
+///
+/// Replace the contract code at the specified address with new code.
+///
+/// # Note
+///
+/// There are a couple of important considerations which must be taken into account when
+/// using this API:
+///
+/// 1. The storage at the code hash will remain untouched. This means that contract developers
+/// must ensure that the storage layout of the new code is compatible with that of the old code.
+///
+/// 2. Contracts using this API can't be assumed as having deterministic addresses. Said another way,
+/// when using this API you lose the guarantee that an address always identifies a specific code hash.
```

Comment:
I don't fully get it, the `AccountId` behind which the contract is will never change, but here it sounds like it could.

Reply:
This whole doc comment is simply copied from the `set_code_hash` impl. 

Anyway I think the comment is accurate, because the `AccountId` is derived from [`hash(deploying_address ++ code_hash ++ salt)`](https://github.com/paritytech/substrate/blob/982f5998c59bd2bd455808345ae1bd2b1767f353/frame/contracts/src/lib.rs#L187). If the `code_hash` of a contract account were immutable you would have the guarantee that you could always derive the `AccountId` from these values. However given the code_hash is mutable, that guarantee no longer exists.

Reply:
FYI I have now just removed the duplicated doc comments, linking to the original `set_code_hash`

Reply:
> I don't fully get it, the `AccountId` behind which the contract is will never change, but here it sounds like it could.

The comment states that the `code_hash` is mutable **not** the contract address. The consequence is that you can't assume that you can derive the contract address of a contract deterministically that uses this function. But not because the address can change. The code hash changes.

Reply:
>2. Contracts using this API can't be assumed as having deterministic addresses. Said another way,

How does this state immutability?

Reply:
Okay, we have to clarify that in the comment. Because this:

>2. Contracts using this API can't be assumed as having deterministic addresses. 

will definitely cause confusion. The misunderstanding here would be the same train of thought I had: that you cannot necessarily rely on the contract always being behind this address after you use `set_code_hash`.